rust/connlib/tunnel/src/device_channel/tun_linux.rs

```diff
@@ -231,8 +231,17 @@ async fn set_iface_config(config: InterfaceConfig, _: Vec<IpAddr>, handle: Handl
         .await;
 
     handle.link().set(index).up().execute().await?;
-
     res_v4.or(res_v6)?;
+
+    // TODO: Having this inside the library is definitely wrong. I think `set_iface_config`
+    // needs to return before `new` returns, so that the `on_tunnel_ready` callback
+    // happens after the IP address and DNS are set up. Then we can call `sd_notify`
```

Comment:
👍

Reply:
Yeah, hopefully soon we can simplify these callbacks and make the clients have a similar eventloop to the gateway which should allow us to selectively integrate systemd only in the linux-client.

Reply:
Hmm actually one quick question on this. I'm a bit confused -- does the `on_tunnel_ready` callback on Linux *not* happen last? On the other platforms that would be the ideal place to notify a hypothetical systemd that's we're up successfully.

Reply:
@jamilbk Right now on Linux `on_tunnel_ready` and `on_set_iface_config` aren't called at all because I'm working around this: https://github.com/firezone/firezone/blob/5f126fa8f299dee777826a08d76c6e4ca294a4cf/rust/connlib/tunnel/src/device_channel/tun_linux.rs#L50

The `Tun` object only gets an `&Callbacks` in its ctor, and then most of the setup work is done async in a worker task that hangs off of `poll_read` https://github.com/firezone/firezone/blob/5f126fa8f299dee777826a08d76c6e4ca294a4cf/rust/connlib/tunnel/src/device_channel/tun_linux.rs#L80

I don't really understand this. But the upshot is that the important state change callbacks can't be fired since I can't figure out how to pass the callbacks ref in there. I just finish the DNS work in this dangling Tokio task. If it fails you get a `tracing` error but `new` has already exited and `on_tunnel_ready` already fired too early, so the client can't actually react to it. https://github.com/firezone/firezone/blob/5f126fa8f299dee777826a08d76c6e4ca294a4cf/rust/connlib/tunnel/src/device_channel/tun_linux.rs#L253

To be blunt I don't understand the architecture of the callbacks. There is tons of platform-specific stuff but you can't call any blocking API because the functions run inside a Tokio context, but you also can't call any async functions because you can't call `block_on` inside a Tokio context. It makes it impossible to do almost anything but pure computation, in the only part of the code where I would think we'd be doing OS-specific I/O...

It sounds like this is fixed in snownet's design (Maybe I should look at it?) but in firezone-tunnel on Windows and now Linux it's bugged me.